File: src/engine/findings.ts
import type { LiftWord, Token } from './types';

const STOPWORDS = new Set(['a', 'an', 'and', 'the', 'of', 'in', 'on', 'to', 'it', 'is', 'was', 'with', 'by', 'for', 'one', 'no', 'not', 'that', 'this', 'has', 'been', 'who', 'at', 'from', 'then', 'too', 'very', 'since', 'there', 'now', 'every', 'keeps', 'says', 'does']);

export interface Findings {
  hourAll: number[];
  hourWin: number[];
  baseline: number;
  lift: LiftWord[];
  loreCorr: number;
}

// #region stage:conclusions
/** What the Console reports: survival by launch hour, lore words with lift, and lore-length correlation. */
export function computeFindings(labelled: readonly Token[]): Findings {
  const hourAll = new Array<number>(24).fill(0);
  const hourWin = new Array<number>(24).fill(0);
  const wordAll = new Map<string, number>();
  const wordWin = new Map<string, number>();
  let pass = 0;
  let sx = 0;
  let sxy = 0;
  let sxx = 0;

  for (const t of labelled) {
    const passed = t.status === 'passed';
    hourAll[t.hour]++;
    if (passed) {
      hourWin[t.hour]++;
      pass++;
    }
    const words = new Set((t.loreRaw ?? t.lore).toLowerCase().match(/[a-z]+/g) ?? []);
    for (const w of words) {
      wordAll.set(w, (wordAll.get(w) ?? 0) + 1);
      if (passed) wordWin.set(w, (wordWin.get(w) ?? 0) + 1);
    }
    const x = words.size;
    sx += x;
    sxx += x * x;
    if (passed) sxy += x;
  }

  const n = labelled.length;
  const baseline = n ? pass / n : 0;
  const lift = [...wordAll]
    .filter(([w, count]) => count >= 60 && !STOPWORDS.has(w))
    .map(([word, count]) => ({ word, n: count, lift: baseline ? (wordWin.get(word) ?? 0) / count / baseline : 0 }))
    .sort((a, b) => b.lift - a.lift)
    .slice(0, 6);

  // Point-biserial correlation between lore length and survival.
  const den = Math.sqrt((n * sxx - sx * sx) * (n * pass - pass * pass));
  const loreCorr = den ? (n * sxy - sx * pass) / den : 0;

  return { hourAll, hourWin, baseline, lift, loreCorr };
}
// #endregion

File: backend/wassily/engine/findings.py
"""What the Console reports: survival by launch hour, lore words with lift, and lore-length correlation."""

from __future__ import annotations

import math
import re
from collections.abc import Iterable
from dataclasses import dataclass, field

from wassily.engine.types import LiftWord, Token

STOPWORDS = frozenset(
    "a an and the of in on to it is was with by for one no not that this has been who at from then too very "
    "since there now every keeps says does".split()
)
MIN_WORD_COUNT = 60
TOP_WORDS = 6
_WORD = re.compile(r"[a-z]+")


@dataclass
class Findings:
    hour_all: list[int] = field(default_factory=lambda: [0] * 24)
    hour_win: list[int] = field(default_factory=lambda: [0] * 24)
    baseline: float = 0.0
    lift: list[LiftWord] = field(default_factory=list)
    lore_corr: float = 0.0


def compute_findings(labelled: Iterable[Token]) -> Findings:
    f = Findings()
    word_all: dict[str, int] = {}
    word_win: dict[str, int] = {}
    n = 0
    passes = 0
    sx = 0.0
    sxy = 0.0
    sxx = 0.0

    for t in labelled:
        n += 1
        passed = t.status == "passed"
        f.hour_all[t.hour] += 1
        if passed:
            f.hour_win[t.hour] += 1
            passes += 1
        text = t.lore_raw if t.lore_raw is not None else t.lore
        words = set(_WORD.findall(text.lower()))
        for w in words:
            word_all[w] = word_all.get(w, 0) + 1
            if passed:
                word_win[w] = word_win.get(w, 0) + 1
        x = len(words)
        sx += x
        sxx += x * x
        if passed:
            sxy += x

    f.baseline = passes / n if n else 0.0
    f.lift = sorted(
        (
            LiftWord(word=w, n=count, lift=(word_win.get(w, 0) / count / f.baseline) if f.baseline else 0.0)
            for w, count in word_all.items()
            if count >= MIN_WORD_COUNT and w not in STOPWORDS
        ),
        key=lambda lw: -lw.lift,
    )[:TOP_WORDS]

    # Point-biserial correlation between lore length and survival.
    den_sq = (n * sxx - sx * sx) * (n * passes - passes * passes)
    f.lore_corr = (n * sxy - sx * passes) / math.sqrt(den_sq) if den_sq > 0 else 0.0
    return f
